fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/OCFLPersistentStorageSession.java

```diff
@@ -166,7 +279,48 @@ public void rollback() throws PersistentStorageException {
             // No changes to rollback
             return;
         }
-        // rollback changes
+
+        if (!state.equals(State.COMMIT_FAILED) && !state.equals(State.PREPARE_FAILRED)) {
+            throw new PersistentStorageException("This session cannot be rolled back in this state: " + state);
+        }
+
+        this.state = State.ROLLING_BACK;
+        LOGGER.info("rolling back...");
+
+        //close any uncommitted sessions
+        final List<OCFLObjectSession> committedSessions = this.sessionsToRollback.stream().map(c -> c.session).collect(Collectors.toList());
+        final List<OCFLObjectSession> uncommittedSessions = new ArrayList<>(this.sessionMap.values());
+        uncommittedSessions.removeAll(committedSessions);
+        for (OCFLObjectSession obj : uncommittedSessions) {
+            obj.close();
+        }
+
+        // rollback committed sessions
+        //for each committed session, rollback if possible
+        final List<String> rollbackFailures = new ArrayList<>(this.sessionsToRollback.size());
+        for (CommittedSession cs : this.sessionsToRollback) {
+            if (cs.option == NEW_VERSION) {
+                //TODO rollback to previous OCFL version
+                //add any failure messages here.
+            } else {
+                rollbackFailures.add(format("%s was already committed to the unversioned head", cs.session));
+            }
+        }
+        //throw an exception if any sessions could not be rolled back.
+        if (rollbackFailures.size() > 0) {
+            state = State.ROLLBACK_FAILED;
+            final StringBuilder builder = new StringBuilder();
+            builder.append("Unable to rollback successfully due to the following reasons: \n");
+            for (String failures : rollbackFailures) {
+                builder.append("\t\t" + failures + "\n");
```

Comment:
Wonder if we should use spaces? I don't see any other uses of tabs in the project, but i also don't know if we have any list structured exception bodies like this.

Reply:
👍